src/NumericalAlgorithms/Interpolation/InterpolationTargetReceiveVars.hpp

```diff
@@ -0,0 +1,188 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/Tags.hpp"
+#include "NumericalAlgorithms/Interpolation/Tags.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// \cond
+namespace intrp {
+template <class Metavariables, size_t VolumeDim>
+struct Interpolator;
+template <typename Metavariables, typename InterpolationTargetTag,
+          size_t VolumeDim, typename Frame>
+class InterpolationTarget;
+namespace Actions {
+template <typename InterpolationTargetTag>
+struct CleanUpInterpolator;
+}  // namespace Actions
+}  // namespace intrp
+/// \endcond
+
+namespace intrp {
+
+namespace InterpolationTarget_detail {
+/// Calls the callback function, tells interpolators to clean up the current
+/// temporal_id, and then if there are more temporal_ids to be interpolated,
+/// starts the next one.
+template <typename InterpolationTargetTag, size_t VolumeDim, typename Frame,
+          typename DbTags, typename Metavariables>
+void callback_and_cleanup(
+    db::DataBox<DbTags>& box,
+    Parallel::ConstGlobalCache<Metavariables>& cache) noexcept {
+  const auto temporal_id =
+      db::get<Tags::TemporalIds<Metavariables>>(box).front();
+
+  // Do the callback
+  const auto& new_box = db::create_from<
+      db::RemoveTags<>, db::AddSimpleTags<>,
+      db::AddComputeTags<
+          typename InterpolationTargetTag::compute_items_on_target>>(box);
+  InterpolationTargetTag::post_interpolation_callback::apply(new_box, cache,
+                                                             temporal_id);
+
+  // Remove current temporal_id.  This is done after the callback
+  // because some callbacks (e.g. the horizon finder) will call
+  // InterpolationTargetTag::compute_target_points again, with the
+  // same temporal_id, to trigger another interpolation (e.g. another
+  // iteration of the horizon finder) and will end up here only after
+  // several interpolations (e.g. after the horizon finder has
+  // converged).
```

Comment:
This comment was intended to explain why `TemporalIds<Metavariables>.pop_front()` is called after `post_interpolation_callback` rather than before `post_interpolation_callback`.  For the current tests, it does not matter if `TemporalIds<Metavariables>.pop_front()` is called before or after `post_interpolation_callback`.  But for horizon finding, `callback_and_cleanup` will be called multiple times for the same `temporal_id` (because `post_interpolation_callback` will end up calling `callback_and_cleanup` down the chain in that case), and we don't want to be popping temporal_ids from the stack for each of these calls.

Is there a better way to word this comment?

Reply:
I don't understand what your intended control flow is.  Will post_interpolation_callback call callback_and_cleanup directly or through the InterpolationTargetReceiveVars action?  In the latter case, I don't think the code runs in the order the comment says it does.  In either case, you're still getting a pop for each call to callback_and_cleanup, so I'm still not sure what your intent is.

Reply:
Ok, I think I now understand your issue with this comment.   And I now agree that callback_and_cleanup will need to be changed more extensively for the case in which post_interpolation_callback calls InterpolationTargetReceiveVars.  So I'll just remove the comment,
perhaps replacing with "This needs to be thought through more carefully when adding the apparent horizon finder".